web3.js/src/connection.js

```diff
@@ -1951,39 +1952,74 @@ export class Connection {
   /**
    * Confirm the transaction identified by the specified signature.
    *
-   * If `confirmations` count is not specified, wait for transaction to be finalized.
-   *
+   * If `commitment` is not specified, default to 'max'.
    */
   async confirmTransaction(
     signature: TransactionSignature,
-    confirmations: ?number,
-  ): Promise<RpcResponseAndContext<SignatureStatus | null>> {
+    commitment: ?Commitment,
+  ): Promise<RpcResponseAndContext<SignatureResult>> {
+    let decodedSignature;
+    try {
+      decodedSignature = bs58.decode(signature);
+    } catch (err) {
+      throw new Error('signature must be base58 encoded: ' + signature);
+    }
+
+    assert(decodedSignature.length === 64, 'signature has invalid length');
+
     const start = Date.now();
-    const WAIT_TIMEOUT_MS = 60 * 1000;
+    const subscriptionCommitment: Commitment = commitment || 'max';
 
-    let statusResponse = await this.getSignatureStatus(signature);
-    for (;;) {
-      const status = statusResponse.value;
-      if (status) {
-        // 'status.confirmations === null' implies that the tx has been finalized
-        if (
-          status.err ||
-          status.confirmations === null ||
-          (typeof confirmations === 'number' &&
-            status.confirmations >= confirmations)
-        ) {
-          break;
-        }
-      } else if (Date.now() - start >= WAIT_TIMEOUT_MS) {
+    let subscriptionId;
+    let response: RpcResponseAndContext<SignatureResult> | null = null;
+    const confirmPromise = new Promise((resolve, reject) => {
+      try {
+        subscriptionId = this.onSignature(
+          signature,
+          (result, context) => {
+            subscriptionId = undefined;
+            response = {
+              context,
+              value: result,
+            };
+            resolve();
+          },
+          subscriptionCommitment,
+        );
+      } catch (err) {
+        reject(err);
+      }
+    });
+
+    let timeoutMs = 60 * 1000;
```

Comment:
Not necessarily a part of this PR but I think we can do better than these timeouts (60s/10s).  The transaction is valid until the blockhash expires, so really that's what we should be waiting for as the time upper bound.

Reply:
Agreed, here's an issue for that: https://github.com/solana-labs/solana/issues/12103